### DISCUSSION TITLE
Help Needed: Handling Multiple Joins in Dynamically Constructed SQLAlchemy Queries

### DISCUSSION THREAD

**Discussion by Huangyh1122** [sqlalchemy/sqlalchemy#11699] [Usage Questions]
Hi SQLAlchemy community,

I'm currently working on a project where I dynamically construct SQL queries using SQLAlchemy based on JSON input. My goal is to build queries step-by-step, handling various SQL operations like SELECT, WHERE, JOIN, ORDER BY, etc.

However, I'm facing an issue when trying to handle multiple joins in the query. Specifically, I'm running into errors with the current implementation of my join logic. Below is the relevant part of my code for handling joins:
```py
if stage['join_type'] == 'left':
    query = query.select_from(local_table_object).outerjoin(foreign_table_object, local_field_object == foreign_field_object)
elif stage['join_type'] == 'right':
    query = query.select_from(foreign_table_object).outerjoin(local_table_object, foreign_field_object == local_field_object)
else:
    query = query.join(foreign_table_object, local_field_object == foreign_field_object)
```

I'm using JSON input to describe the query stages and construct the query step-by-step. Here's a simplified example of the JSON input structure for a query with multiple joins:
```json
[
    {
        "type": "select",
        "fields": [...],
        "calculated": [...]
    },
    {
        "type": "join",
        "join_type": "left",
        "local_field": "table1.id",
        "foreign_field": "table2.table1_id"
    },
    {
        "type": "join",
        "join_type": "inner",
        "local_field": "table2.id",
        "foreign_field": "table3.table2_id"
    },
    ...
]
```
The problem arises because SQL standard queries have the concept of a main table when performing joins, but SQLAlchemy does not have this concept directly. As a result, handling multiple joins where the main table can change is challenging.

Could anyone advise on the correct way to handle multiple joins dynamically in SQLAlchemy? How can I manage the concept of a main table or change the main table appropriately while constructing the query step-by-step?

**Comment by CaselIT** [comment]
Hi,

I don't think you need to set `select_from(local_table_object)`. Assuming the main table is the one you select from originally, just provide the table you want to join with and the join condition

**Reply by Huangyh1122** [reply]
But what about multi join, or the main table is different with the one I select from originally

**Reply by CaselIT** [reply]
you can use `join()` directly, then join that to the main query.
That allows you to render something like
```sql
select ...
from a
join (b join c on ...) on ...
```

example:

```py
sa.select('*').select_from(sa.table('a')).join(
  sa.join(sa.table('b'), sa.table('c'), sa.column('x') == sa.column('y')),
  sa.column('w') == sa.column('z')
)
